pkg/kubernetes/status/health.go

```diff
@@ -51,95 +47,108 @@ func IsDeleted(client client.Client, discovery discovery.CachedDiscoveryInterfac
 	return nil
 }
 
-// IsHealthy returns whether an object is healthy. Must be implemented for each type.
-func IsHealthy(obj runtime.Object) error {
+// IsTerminallyFailed returns true if a resource will never become healthy anymore
+func IsTerminallyFailed(obj runtime.Object) (bool, string, error) {
 	if obj == nil {
-		return nil
+		return true, "", nil
 	}
-	unstructMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(obj)
-	if err != nil {
-		return err
+
+	switch obj := obj.(type) {
+	case *batchv1.Job:
+		return isJobTerminallyFailed(obj)
 	}
+	return false, "", nil
+}
 
-	objUnstructured := &unstructured.Unstructured{Object: unstructMap}
+// IsHealthy returns whether an object is healthy and a corresponding message
+// Must be implemented for each type.
+func IsHealthy(obj runtime.Object) (bool, string, error) {
```

Comment:
could we add more godoc here...  developer less familiar needs to read all code to understand you signature.  rules would also be helpful... there is a lot of complexity here... rules like:  is it possible for bool to be true with an error?   I would think not.   Is there a case of a string being empty other than error?

IMO it would also be helpful to provide name variables here which would help with context.  `(healthy bool, msg string, err error) ` is more readable to me... otherwise it is a hunt for what is the string.

Reply:
Let's also mention that only certain common objects are checked for health. Unknown objects will be considered healthy by default.

Reply:
👍